### DISCUSSION TITLE
Prebuilt musl pdfium?

### DISCUSSION THREAD

**Discussion by tuupke** [libvips/libvips#3645] [Q&A]
While investigating whether I could get libvips with pdfium support up and running in Alpine, I saw that there is now [prebuilt pdfium for musl](https://github.com/bblanchon/pdfium-binaries/commit/925b894af214e655f55d76e7c5b792802cb41a18).

With good hope - and a good dose of naivete - I set out to write a dockerfile (added below) cobbled together from the [examples by jcupitt](https://github.com/jcupitt/docker-builds). Too much naivete as it turns out, because my attempt failed. Hopefully due to my lack of experience working with c and c++.

Everything appears to be working fine up-to `meson build`. Then it suddenly stops when it is - I think, again naivete - linking.
[This](https://github.com/libvips/libvips/files/12501533/meson.compile.log) is the log of the `meson compile` step.

Having looked for issues in both libvips and pdfium-binaries repo's I could not find a similar issue or another person's attempt on getting this working with Alpine. I could only find one [issue](https://github.com/bblanchon/pdfium-binaries/issues/52) (that pointed me to the commit listed above) but looking into that did not point me towards what to do.

Am I completely misunderstanding what the musl build of pdfium actually does, or is there a mistake here that prevents this from working?

<details>
<summary>Dockerfile</summary>

```
FROM alpine:3

ARG PREFIX=/usr/local/vips

USER root
RUN apk add \
    build-base \
    autoconf \
    automake \
    libtool \
    meson \
    cmake \
    bc \
    zlib-dev \
    libxml2-dev \
    jpeg-dev \
    openjpeg-dev \
    tiff-dev \
    glib-dev \
    gdk-pixbuf-dev \
    sqlite-dev \
    cgif-dev \
    libjpeg-turbo-dev \
    libexif-dev \
    lcms2-dev \
    fftw-dev \
    giflib-dev \
    libpng-dev \
    libwebp-dev \
    orc-dev \
    librsvg-dev \
    libspng-dev \
    libgsf-dev \
    openexr-dev \
    libheif-dev \
    fftw-dev \
    gtk-doc \
    tar \
    xz \
    curl

RUN mkdir -p ${PREFIX}/lib/pkgconfig

WORKDIR ${PREFIX}

# Checkout VIPS dir
ARG VIPS_VERSION=8.14.4
ARG VIPS_URL=https://github.com/libvips/libvips/releases/download
RUN curl -L ${VIPS_URL}/v${VIPS_VERSION}/vips-${VIPS_VERSION}.tar.xz | tar vxfJ - --strip-components=1


# Add pdfium
ARG PDFIUM_VERSION=5854
ARG PDFIUM_URL=https://github.com/bblanchon/pdfium-binaries/releases/download/chromium
RUN curl -L ${PDFIUM_URL}%2F${PDFIUM_VERSION}/pdfium-linux-musl-x64.tgz | tar xfz -


COPY <<EOF /usr/lib/pkgconfig/pdfium.pc
prefix=$PREFIX
exec_prefix=\${prefix}
libdir=\${exec_prefix}/lib
includedir=\${prefix}/include
Name: pdfium
Description: pdfium
Version: ${PDFIUM_VERSION}
Requires:
Libs: -L\${libdir} -lpdfium
Cflags: -I\${includedir}
EOF

RUN meson setup build -Dintrospection=false -Dmodules=disabled -Dexamples=false -Ddebug=false -Dcgif=enabled -Dpdfium=enabled
RUN    cd build; \
    meson compile; \
    meson install;

```

</details>

*edited to add the log

**Comment by jcupitt** [comment]
Hi, looks like you're missing the right `libstdc++` version. The musl binary is built against 3.4.11, but you have some other version on your system.

C++ is especially tricky to link as the language does not define an ABI. You have to build the entire project with exactly the same compiler, linker, flags, libraries ... everything.

**Reply by tuupke** [reply]
Ok, so the only way forward is to take the build steps as executed in the [prebuilt pdfium](https://github.com/bblanchon/pdfium-binaries) repo and do them within alpine. Giving that a crack then :sweat_smile: Thanks!
